pdf2htmlEX/pdf2htmlEX.1.in
.TH pdf2htmlEX 1 "pdf2htmlEX @PDF2HTMLEX_VERSION@"
.SH NAME
.PP
.nf
  pdf2htmlEX \- converts PDF to HTML without losing text and format.
.fi

.SH USAGE
.PP
.nf
  pdf2htmlEX [options] <input\-filename> [<output\-filename>]
.fi

.SH DESCRIPTION
.PP
pdf2htmlEX is a utility that converts PDF files to HTML files.

pdf2htmlEX tries its best to render the PDF precisely, maintain proper styling, while retaining text and optimizing for Web.

Fonts are extracted form PDF and then embedded into HTML, text in the converted HTML file is usually selectable and copyable. 

Other objects are rendered as images and also embedded.

.SH OPTIONS

.SS Pages

.TP
.B \-f, \-\-first\-page <num> (Default: 1)
Specify the first page to process

.TP
.B \-l, \-\-last\-page <num> (Default: last page)
Specify the last page to process

.SS Dimensions

.TP
.B \-\-zoom <ratio>, \-\-fit\-width <width>, \-\-fit\-height <height>
\-\-zoom specifies the zoom factor directly; \-\-fit\-width/height specifies the maximum width/height of a page, the values are in pixels.

If multiple values are specified, the minimum one will be used.

If none is specified, pages will be rendered as 72DPI.

.TP
.B \-\-use\-cropbox <0|1> (Default: 1)
Use CropBox instead of MediaBox for output.

.TP
.B \-\-hdpi <dpi>, \-\-vdpi <dpi> (Default: 144)
Specify the horizontal and vertical DPI for images


.SS Output 

.B \-\-embed <string>
.br
.B \-\-embed\-css <0|1> (Default: 1)
.br
.B \-\-embed\-font <0|1> (Default: 1)
.br
.B \-\-embed\-image <0|1> (Default: 1)
.br
.B \-\-embed\-javascript <0|1> (Default: 1)
.br
.B \-\-embed\-outline <0|1> (Default: 1)
.RS
Specify which elements should be embedded into the output HTML file.

If switched off, separated files will be generated along with the HTML file for the corresponding elements.

\-\-embed accepts a string as argument. Each letter of the string must be one of `cCfFiIjJoO`, which corresponds
to one of the \-\-embed\-*** switches. Lower case letters for 0 and upper case letters for 1. For example,
`\-\-embed cFIJo` means to embed everything but CSS files and outlines.
.RE
.TP
.B \-\-split\-pages <0|1> (Default: 0)
If turned on, the content of each page is stored in a separated file.

This switch is useful if you want pages to be loaded separately & dynamically \-\- a supporting server might be necessary.

Also see \-\-page\-filename.

.TP
.B \-\-dest\-dir <dir> (Default: .)
Specify destination folder.

.TP
.B \-\-css\-filename <filename> (Default: <none>)
Specify the filename of the generated css file, if not embedded.

If it's empty, the file name will be determined automatically.

.TP
.B \-\-page\-filename <filename> (Default: <none>)
Specify the filename template for pages when \-\-split\-pages is 1

A %d placeholder may be included in `filename` to indicate where the page number should be placed. The placeholder supports a limited subset of normal numerical placeholders, including specified width and zero padding.

If `filename` does not contain a placeholder for the page number, the page number will be inserted directly before the file extension. If the filename does not have an extension, the page number will be placed at the end of the file name.

If \-\-page\-filename is not specified, <input\-filename> will be used for the output filename, replacing the extension with .page and adding the page number directly before the extension.

.B Examples

.B pdf2htmlEX \-\-split\-pages 1 foo.pdf

  Yields page files foo1.page, foo2.page, etc.

.B pdf2htmlEX \-\-split\-pages 1 foo.pdf \-\-page\-filename bar.baz

  Yields page files bar1.baz, bar2.baz, etc.

.B pdf2htmlEX \-\-split\-pages 1 foo.pdf \-\-page\-filename page%dbar.baz

  Yields page files page1bar.baz, page2bar.baz, etc.

.B pdf2htmlEX \-\-split\-pages 1 foo.pdf \-\-page\-filename bar%03d.baz

  Yields page files bar001.baz, bar002.baz, etc.

.TP
.B \-\-outline\-filename <filename> (Default: <none>)
Specify the filename of the generated outline file, if not embedded. 

If it's empty, the file name will be determined automatically.

.TP
.B \-\-process\-nontext <0|1> (Default: 1)
Whether to process non\-text objects (as images)

.TP
.B \-\-process\-outline <0|1> (Default: 1)
Whether to show outline in the generated HTML

.TP
.B \-\-process-annotation <0|1> (Default: 0)
Whether to show annotation in the generated HTML

.TP
.B \-\-process-form <0|1> (Default: 0)
Whether to include text fields and radio buttons in the generated HTML

.TP
.B \-\-printing <0|1> (Default: 1)
Enable printing support. Disabling this option may reduce the size of CSS.

.TP
.B \-\-fallback <0|1> (Default: 0)
Output in fallback mode, for better accuracy and browser compatibility, but the size becomes larger.

.TP
.B \-\-tmp\-file\-size\-limit <limit> (Default: \-1)
This limits the total size (in KB) of the temporary files which will also limit the total size of the output file.
This is an estimate and it will stop after a page, once the total temporary files size is greater than this number.

\-1 means no limit and is the default.


.SS Fonts

.TP
.B \-\-embed\-external\-font <0|1> (Default: 1)
Specify whether the local matched fonts, for fonts not embedded in PDF, should be embedded into HTML. 

If this switch is off, only font names are exported such that web browsers may try to find proper fonts themselves, and that might cause issues about incorrect font metrics.

.TP
.B \-\-font\-format <format> (Default: woff)
Specify the format of fonts extracted from the PDF file.

.TP
.B \-\-decompose\-ligature <0|1> (Default: 0)
Decompose ligatures. For example 'fi' \-> 'f''i'.

.TP
.B \-\-turn\-off\-ligatures <0|1> (Default: 0)
Explicitly tell browsers not to use ligatures.

.TP
.B \-\-auto\-hint <0|1> (Default: 0)
If set to 1, hints will be generated for the fonts using FontForge. 

This may be preceded by \-\-external\-hint\-tool.

.TP
.B \-\-external\-hint\-tool <tool> (Default: <none>)
If specified, the tool will be called in order to enhanced hinting for fonts, this will precede \-\-auto\-hint.

The tool will be called as '<tool> <in.suffix> <out.suffix>', where suffix will be the same as specified for \-\-font\-format.

.TP
.B \-\-stretch\-narrow\-glyph <0|1> (Default: 0)
If set to 1, glyphs narrower than described in PDF will be stretched; otherwise space will be padded to the right of the glyphs

.TP
.B \-\-squeeze\-wide\-glyph <0|1> (Default: 1)
If set to 1, glyphs wider than described in PDF will be squeezed; otherwise it will be truncated.

.TP
.B \-\-override\-fstype <0|1> (Default: 0)
Clear the fstype bits in TTF/OTF fonts. 

Turn this on if Internet Explorer complains about 'Permission must be Installable' AND you have permission to do so.

.TP
.B \-\-process\-type3 <0|1> (Default: 0)
If turned on, pdf2htmlEX will try to convert Type 3 fonts such that text can be rendered natively in HTML.
Otherwise all text with Type 3 fonts will be rendered as image.

This feature is highly experimental.

.SS Text

.TP
.B \-\-heps <len>, \-\-veps <len> (Default: 1)
Specify the maximum tolerable horizontal/vertical offset (in pixels).

pdf2htmlEX would try to optimize the generated HTML file moving Text within this distance.

.TP
.B \-\-space\-threshold <ratio> (Default: 0.125)
pdf2htmlEX would insert a whitespace character ' ' if the distance between two consecutive letters in the same line is wider than ratio * font_size.

.TP
.B \-\-font\-size\-multiplier <ratio> (Default: 4.0)
Many web browsers limit the minimum font size, and many would round the given font size, which results in incorrect rendering.

Specify a ratio greater than 1 would resolve this issue, however it might freeze some browsers.

For some versions of Firefox, however, there will be a problem when the font size is too large, in which case a smaller value should be specified here.

.TP
.B \-\-space\-as\-offset <0|1> (Default: 0)
If set to 1, space characters will be treated as offsets, which allows a better optimization. 

For PDF files with bad encodings, turning on this option may cause losing characters.

.TP
.B \-\-tounicode <\-1|0|1> (Default: 0)
A ToUnicode map may be provided for each font in PDF which indicates the 'meaning' of the characters. However often there is better "ToUnicode" info in Type 0/1 fonts, and sometimes the ToUnicode map provided is wrong. 
If this value is set to 1, the ToUnicode Map is always applied, if provided in PDF, and characters may not render correctly in HTML if there are collisions.

If set to \-1, a customized map is used such that rendering will be correct in HTML (visually the same), but you may not get correct characters by select & copy & paste.

If set to 0, pdf2htmlEX would try its best to balance the two methods above.

.TP
.B \-\-optimize\-text <0|1> (Default: 0)
If set to 1, pdf2htmlEX will try to reduce the number of HTML elements used for text. Turn it off if anything goes wrong.

.TP
.B --correct-text-visibility <0|1|2> (Default: 1)
0 : Do not do visibility calculations for text
1 : Text fully occluded will be drawn in the background layer
2 : Text partially occluded will be drawn in the background layer (more false positives than option "1")

.SS Background Image

.TP
.B \-\-bg\-format <format> (Default: png)
Specify the background image format. Run `pdf2htmlEX \-v` to check all supported formats.

.TP
.B \-\-svg\-node\-count\-limit <limit> (Default: -1)
If node count in a svg background image exceeds this limit, fall back this page to bitmap background; negative value means no limit.
This option is only useful when '\-\-bg\-format svg' is specified. Note that node count in svg is just calculated approximately.

.TP
.B \-\-svg\-embed\-bitmap <0|1> (Default: 1)
Whether embed bitmaps in svg background image. 1: embed bitmaps in svg background; 0: dump bitmaps to external files if possible.

This option is only useful when '\-\-bg\-format svg' is specified and '\-\-embed\-image' is off.

Currently, RGB or Gray JPEG bitmaps in a PDF can be dumped, while those in other formats or colorspaces are still embedded.
If bitmaps are not dumped as expected, try pre-processing your PDF by ghostscript or acrobat and make sure bitmaps in it are converted to RGB/Gray JPEG format. See the project wiki for more details.

.SS PDF Protection

.TP
.B \-o, \-\-owner\-password <password>
Specify owner password

.TP
.B \-u, \-\-user\-password <password>
Specify user password

.TP
.B \-\-no\-drm <0|1> (Default: 0)
Override document DRM settings

Turn this on only when you have permission.

.SS Misc.

.TP
.B \-\-clean\-tmp <0|1> (Default: 1)
If switched off, intermediate files won't be cleaned in the end.

.TP
.B \-\-data\-dir <dir> (Default: @CMAKE_INSTALL_PREFIX@/share/pdf2htmlEX)
Specify the folder holding the manifest and other files (see below for the manifest file)`


debian/changelog
pdf2htmlex (0.18.7-1~git20200531r3fcd0-0ubuntu1) UNRELEASED; urgency=medium

  * test packaging

 -- Stephen Gaito <dev@pDev>  Sun, 31 May 2020 13:38:59 +0000

pdf2htmlex (0.18.7-1~git20190927rb69b8-0ubuntu1) disco; urgency=low

  * Package for poppler-0.81.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Fri, 27 Sep 2019 12:55:39 +0000

pdf2htmlex (0.18.6-1~git20190927r583b1-0ubuntu1) disco; urgency=low

  * Package for poppler-0.80.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Fri, 27 Sep 2019 11:58:57 +0000

pdf2htmlex (0.18.5-1~git20190927rdb4cc-0ubuntu1) disco; urgency=low

  * Package for poppler-0.79.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Fri, 27 Sep 2019 09:12:53 +0000

pdf2htmlex (0.18.4-1~git20190927r34bac-0ubuntu1) disco; urgency=low

  * Package for poppler-0.78.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Fri, 27 Sep 2019 07:46:26 +0000

pdf2htmlex (0.18.3-1~git20190926r7ed19-0ubuntu1) disco; urgency=low

  * Package for poppler-0.77.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Thu, 26 Sep 2019 17:56:51 +0000

pdf2htmlex (0.18.2-1~git20190926ra37f7-0ubuntu1) disco; urgency=low

  * Package for poppler-0.76.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Thu, 26 Sep 2019 15:29:27 +0000

pdf2htmlex (0.18.1-1~git20190926r11e2a-0ubuntu1) disco; urgency=low

  * Package for poppler-0.75.0

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Thu, 26 Sep 2019 10:59:59 +0000

pdf2htmlex (0.18.0-1~git20190924r37921-0ubuntu1) disco; urgency=medium

  * Package for Ubuntu 19.04 (poppler-0.74.0)

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Tue, 24 Sep 2019 14:56:03 +0000

pdf2htmlex (0.17.0-1~git20190924r37921-0ubuntu1) cosmic; urgency=low

  * Package for 18.10 (poppler-0.68.0)

 -- Stephen Gaito <stephen@perceptisys.co.uk>  Tue, 24 Sep 2019 10:32:08 +0000

pdf2htmlex (0.16.0-1~git20190920r21a9f-0ubuntu1) bionic; urgency=low

  * Package for 18.04

 -- Trent Petersen <trentpetersen.523@gmail.com>  Thu, 19 Sep 2019 10:33:35 -0500

pdf2htmlex (0.11-1~git201311150048r23755-0ubuntu1) saucy; urgency=low

  * Fix packaging

 -- WANG Lu <coolwanglu@gmail.com>  Fri, 15 Nov 2013 00:48:06 +0800

pdf2htmlex (0.11-1~git201311042119refddc-0ubuntu1) saucy; urgency=low

  * Packaging for 13.10

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 04 Nov 2013 21:19:41 +0800

pdf2htmlex (0.11-1~git201310172203re1b11-0ubuntu1) raring; urgency=low

  * Fix typo in dependencies

 -- WANG Lu <coolwanglu@gmail.com>  Thu, 17 Oct 2013 22:03:42 +0800

pdf2htmlex (0.10-1~git201310171220rc344e-0ubuntu1) raring; urgency=low

  * Fix build

 -- WANG Lu <coolwanglu@gmail.com>  Thu, 17 Oct 2013 12:20:59 +0800

pdf2htmlex (0.10-1~git201310171209rdc970-0ubuntu1) raring; urgency=low

  * v0.10 released, see Changelog

 -- WANG Lu <coolwanglu@gmail.com>  Thu, 17 Oct 2013 12:09:01 +0800

pdf2htmlex (0.9-1~git201309161133rd60eb-0ubuntu1) raring; urgency=low

  * fix cmake test for packaging

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 16 Sep 2013 11:33:11 +0800

pdf2htmlex (0.9-1~git201309161021rf2993-0ubuntu1) raring; urgency=low

  * v0.9 released, see Changelog

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 16 Sep 2013 10:21:05 +0800

pdf2htmlex (0.9-1~git201305291247rf655a-0ubuntu1) raring; urgency=low

  * Regular upload

 -- WANG Lu <coolwanglu@gmail.com>  Wed, 29 May 2013 12:47:16 +0800

pdf2htmlex (0.8-1~git201305051429r92eab-0ubuntu1) raring; urgency=low

  * v0.8 released, see Changelog

 -- WANG Lu <coolwanglu@gmail.com>  Sun, 05 May 2013 14:29:07 +0800

pdf2htmlex (0.8-1~git201304281550rd5f9c-0ubuntu1) raring; urgency=low

  * Test packaging for multiple distributions
  * Fix crashing with --split-pages

 -- WANG Lu <coolwanglu@gmail.com>  Sun, 28 Apr 2013 15:39:34 +0800

pdf2htmlex (0.8-1~git201304151420r1da9b-0ubuntu1) quantal; urgency=low

  * HTML optimization
  * New options: --optimize-text, --fallabck
  * See more in Changelog

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 15 Apr 2013 14:20:22 +0800

pdf2htmlex (0.8-1~git201303011406r3bc73-0ubuntu1) quantal; urgency=low

  * Experimental printing support
  * New version

 -- WANG Lu <coolwanglu@gmail.com>  Fri, 01 Mar 2013 14:06:42 +0800

pdf2htmlex (0.7-1~git201302282259r3bc73-0ubuntu1) quantal; urgency=low

  * suggests ttfautohint

 -- WANG Lu <coolwanglu@gmail.com>  Thu, 28 Feb 2013 22:59:45 +0800

pdf2htmlex (0.7-1~git201302271054r3bc73-0ubuntu1) precise; urgency=low

  * Packaging for 12.04

 -- WANG Lu <coolwanglu@gmail.com>  Wed, 27 Feb 2013 10:54:12 +0800

pdf2htmlex (0.7-1~git201302271053r3bc73-0ubuntu1) quantal; urgency=low

  * Fix packpage dependency

 -- WANG Lu <coolwanglu@gmail.com>  Wed, 27 Feb 2013 10:52:17 +0800

pdf2htmlex (0.7-1~git201302270111r3bc73-0ubuntu1) quantal; urgency=low

  * Regular upload

 -- WANG Lu <coolwanglu@gmail.com>  Wed, 27 Feb 2013 01:11:56 +0800

pdf2htmlex (0.7-1~git201302182049r3bc73-0ubuntu1) quantal; urgency=low

  * New parameter: process-outline

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 18 Feb 2013 20:49:57 +0800

pdf2htmlex (0.7-1~git201301292229r2595c-0ubuntu1) quantal; urgency=low

  * Fixed a CSS issue

 -- WANG Lu <coolwanglu@gmail.com>  Tue, 29 Jan 2013 22:29:21 +0800

pdf2htmlex (0.7-1~git201301282229r2595c-0ubuntu1) quantal; urgency=low

  * Process PDF Outline

 -- WANG Lu <coolwanglu@gmail.com>  Mon, 28 Jan 2013 22:29:35 +0800


pdf2htmlEX/src/pdf2htmlEX.cc
// pdf2htmlEX.cc
//
// Copyright (C) 2012-2015 Lu Wang <coolwanglu@gmail.com>

#include <cstdio>
#include <cstdlib>
#include <cstddef>
#include <cstring>
#include <ctime>
#include <string>
#include <limits>
#include <iostream>
#include <memory>
#include <errno.h>

#include <getopt.h>

#include <poppler-config.h>
#include <goo/GooString.h>

#include <Object.h>
#include <PDFDoc.h>
#include <PDFDocFactory.h>
#include <GlobalParams.h>

#include "pdf2htmlEX-config.h"

#include "util/SignalHandler.h"

#if ENABLE_SVG
#include <cairo.h>
#endif

#include "ArgParser.h"
#include "Param.h"
#include "HTMLRenderer/HTMLRenderer.h"

#include "util/path.h"
#include "util/ffw.h"

#ifdef __MINGW32__
#include "util/mingw.h"
#endif

using namespace std;
using namespace pdf2htmlEX;

Param param;
ArgParser argparser;

void show_usage_and_exit(const char * dummy = nullptr)
{
    cerr << "Usage: pdf2htmlEX [options] <input.pdf> [<output.html>]" << endl;
    argparser.show_usage(cerr);
    exit(EXIT_FAILURE);
}

void show_version_and_exit(const char * dummy = nullptr)
{
    const FFWVersionInfo* ffwVersionInfo = ffw_get_version_info();

    cerr << "pdf2htmlEX version " << PDF2HTMLEX_VERSION << endl;
    cerr << "Copyright 2012-2015 Lu Wang <coolwanglu@gmail.com> and other contributors" << endl;
    cerr << "Libraries: " << endl;
    cerr << "  poppler " << POPPLER_VERSION << endl;
    cerr << "  libfontforge (date) " << ffwVersionInfo->versionDate << endl;
#if ENABLE_SVG
    cerr << "  cairo " << cairo_version_string() << endl;
#endif
    cerr << "Default data-dir: " << param.data_dir << endl;
    cerr << "Poppler data-dir: " << param.poppler_data_dir << endl;
    cerr << "Supported image format:";
#ifdef ENABLE_LIBPNG
    cerr << " png";
#endif
#ifdef ENABLE_LIBJPEG
    cerr << " jpg";
#endif
#if ENABLE_SVG
    cerr << " svg";
#endif
    cerr << endl;

    cerr << endl;
    exit(EXIT_SUCCESS);
}

void embed_parser (const char * str)
{
    while(true)
    {
        switch(*str)
        {
            case '\0': return; break;
            case 'c': param.embed_css = 0; break;
            case 'C': param.embed_css = 1; break;
            case 'f': param.embed_font = 0; break;
            case 'F': param.embed_font = 1; break;
            case 'i': param.embed_image = 0; break;
            case 'I': param.embed_image = 1; break;
            case 'j': param.embed_javascript = 0; break;
            case 'J': param.embed_javascript = 1; break;
            case 'o': param.embed_outline = 0; break;
            case 'O': param.embed_outline = 1; break;
            default:
                cerr << "Unknown character `" << (*str) << "` for --embed" << endl;
                break;
        }
        ++ str;
    }
}

void prepare_directories()
{
    std::string tmp_dir = param.tmp_dir + "/pdf2htmlEX-XXXXXX";

    errno = 0;

    unique_ptr<char[]> pBuf(new char[tmp_dir.size() + 1]);
    strcpy(pBuf.get(), tmp_dir.c_str());
    auto p = mkdtemp(pBuf.get());
    if(p == nullptr)
    {
        const char * errmsg = strerror(errno);
        if(!errmsg)
        {
            errmsg = "unknown error";
        }
        cerr << "Cannot create temp directory: " << errmsg << endl;
        exit(EXIT_FAILURE);
    }
    param.tmp_dir = pBuf.get();
}

void parse_options (int argc, char **argv)
{
    argparser
        // pages
        .add("first-page,f", &param.first_page, 1, "first page to convert")
        .add("last-page,l", &param.last_page, numeric_limits<int>::max(), "last page to convert")

        // dimensions
        .add("zoom", &param.zoom, 0, "zoom ratio", true)
        .add("fit-width", &param.fit_width, 0, "fit width to <fp> pixels", true)
        .add("fit-height", &param.fit_height, 0, "fit height to <fp> pixels", true)
        .add("use-cropbox", &param.use_cropbox, 1, "use CropBox instead of MediaBox")
        .add("dpi", &param.desired_dpi, 144.0, "Resolution for graphics in DPI")

        // output files
        .add("embed", "specify which elements should be embedded into output", embed_parser, true)
        .add("embed-css", &param.embed_css, 1, "embed CSS files into output")
        .add("embed-font", &param.embed_font, 1, "embed font files into output")
        .add("embed-image", &param.embed_image, 1, "embed image files into output")
        .add("embed-javascript", &param.embed_javascript, 1, "embed JavaScript files into output")
        .add("embed-outline", &param.embed_outline, 1, "embed outlines into output")
        .add("split-pages", &param.split_pages, 0, "split pages into separate files")
        .add("dest-dir", &param.dest_dir, ".", "specify destination directory")
        .add("css-filename", &param.css_filename, "", "filename of the generated css file")
        .add("page-filename", &param.page_filename, "", "filename template for split pages ")
        .add("outline-filename", &param.outline_filename, "", "filename of the generated outline file")
        .add("process-nontext", &param.process_nontext, 1, "render graphics in addition to text")
        .add("process-outline", &param.process_outline, 1, "show outline in HTML")
        .add("process-annotation", &param.process_annotation, 0, "show annotation in HTML")
        .add("process-form", &param.process_form, 0, "include text fields and radio buttons")
        .add("printing", &param.printing, 1, "enable printing support")
        .add("fallback", &param.fallback, 0, "output in fallback mode")
        .add("tmp-file-size-limit", &param.tmp_file_size_limit, -1, "Maximum size (in KB) used by temporary files, -1 for no limit")

        // fonts
        .add("embed-external-font", &param.embed_external_font, 1, "embed local match for external fonts")
        .add("font-format", &param.font_format, "woff", "suffix for embedded font files (ttf,otf,woff,svg)")
        .add("decompose-ligature", &param.decompose_ligature, 0, "decompose ligatures, such as \uFB01 -> fi")
        .add("turn-off-ligatures", &param.turn_off_ligatures, 0, "explicitly tell browsers not to use ligatures")
        .add("auto-hint", &param.auto_hint, 0, "use fontforge autohint on fonts without hints")
        .add("external-hint-tool", &param.external_hint_tool, "", "external tool for hinting fonts (overrides --auto-hint)")
        .add("stretch-narrow-glyph", &param.stretch_narrow_glyph, 0, "stretch narrow glyphs instead of padding them")
        .add("squeeze-wide-glyph", &param.squeeze_wide_glyph, 1, "shrink wide glyphs instead of truncating them")
        .add("override-fstype", &param.override_fstype, 0, "clear the fstype bits in TTF/OTF fonts")
        .add("process-type3", &param.process_type3, 0, "convert Type 3 fonts for web (experimental)")

        // text
        .add("heps", &param.h_eps, 1.0, "horizontal threshold for merging text, in pixels")
        .add("veps", &param.v_eps, 1.0, "vertical threshold for merging text, in pixels")
        .add("space-threshold", &param.space_threshold, (1.0/8), "word break threshold (threshold * em)")
        .add("font-size-multiplier", &param.font_size_multiplier, 4.0, "a value greater than 1 increases the rendering accuracy")
        .add("space-as-offset", &param.space_as_offset, 0, "treat space characters as offsets")
        .add("tounicode", &param.tounicode, 0, "how to handle ToUnicode CMaps (0=auto, 1=force, -1=ignore)")
        .add("optimize-text", &param.optimize_text, 0, "try to reduce the number of HTML elements used for text")
        .add("correct-text-visibility", &param.correct_text_visibility, 1, "0: Don't do text visibility checks. 1: Fully occluded text handled. 2: Partially occluded text handled")
        .add("covered-text-dpi", &param.text_dpi, 300, "Rendering DPI to use if correct-text-visibility == 2 and there is partially covered text on the page")

        // background image
        .add("bg-format", &param.bg_format, "png", "specify background image format")
        .add("svg-node-count-limit", &param.svg_node_count_limit, -1, "if node count in a svg background image exceeds this limit,"
                " fall back this page to bitmap background; negative value means no limit")
        .add("svg-embed-bitmap", &param.svg_embed_bitmap, 1, "1: embed bitmaps in svg background; 0: dump bitmaps to external files if possible")

        // encryption
        .add("owner-password,o", &param.owner_password, "", "owner password (for encrypted files)", true)
        .add("user-password,u", &param.user_password, "", "user password (for encrypted files)", true)
        .add("no-drm", &param.no_drm, 0, "override document DRM settings")

        // misc.
        .add("clean-tmp", &param.clean_tmp, 1, "remove temporary files after conversion")
        .add("tmp-dir", &param.tmp_dir, param.tmp_dir, "specify the location of temporary directory")
        .add("data-dir", &param.data_dir, param.data_dir, "specify data directory")
        .add("poppler-data-dir", &param.poppler_data_dir, param.poppler_data_dir, "specify poppler data directory")
        .add("debug", &param.debug, 0, "print debugging information")
        .add("proof", &param.proof, 0, "texts are drawn on both text layer and background for proof")
        .add("quiet", &param.quiet, 0, "perform operations quietly")

        // meta
        .add("version,v", "print copyright and version info", &show_version_and_exit)
        .add("help,h", "print usage information", &show_usage_and_exit)

        .add("", &param.input_filename, "", "")
        .add("", &param.output_filename, "", "")
        ;

    try
    {
        argparser.parse(argc, argv);
    }
    catch(const char * s)
    {
        // if s == "", getopt_long would have printed the error message
        if(s && s[0])
        {
            cerr << "Error when parsing the arguments:" << endl;
            cerr << s << endl;
        }
        exit(EXIT_FAILURE);
    }
    catch(const std::string & s)
    {
        // if s == "", getopt_long would have printed the error message
        if(s != "")
        {
            cerr << "Error when parsing the arguments:" << endl;
            cerr << s << endl;
        }
        exit(EXIT_FAILURE);
    }
}

void check_param()
{
    if (param.input_filename == "")
    {
        show_usage_and_exit();
    }

    if(param.output_filename.empty())
    {
        const string s = get_filename(param.input_filename);
        if(get_suffix(param.input_filename) == ".pdf")
        {
            param.output_filename = s.substr(0, s.size() - 4) + ".html";
        }
        else
        {
            param.output_filename = s + ".html";
        }
    }

    if(param.page_filename.empty())
    {
        const string s = get_filename(param.input_filename);
        if(get_suffix(param.input_filename) == ".pdf")
        {
            param.page_filename = s.substr(0, s.size() - 4) + "%d.page";
        }
        else
        {
            param.page_filename = s + "%d.page";
        }
        sanitize_filename(param.page_filename);
    }

    else
    {
        // Need to make sure we have a page number placeholder in the filename
        if(!sanitize_filename(param.page_filename))
        {
            // Inject the placeholder just before the file extension
            const string suffix = get_suffix(param.page_filename);
            param.page_filename = param.page_filename.substr(0, param.page_filename.size() - suffix.size()) + "%d" + suffix;
            sanitize_filename(param.page_filename);
        }
    }
    if(param.css_filename.empty())
    {
        const string s = get_filename(param.input_filename);

        if(get_suffix(param.input_filename) == ".pdf")
        {
            param.css_filename = s.substr(0, s.size() - 4) + ".css";
        }
        else
        {
            param.css_filename = s + ".css";
        }
    }
    if(param.outline_filename.empty())
    {
        const string s = get_filename(param.input_filename);

        if(get_suffix(param.input_filename) == ".pdf")
        {
            param.outline_filename = s.substr(0, s.size() - 4) + ".outline";
        }
        else
        {
            if(!param.split_pages)
                param.outline_filename = s + ".outline";
        }
    }

    if(false) { }
#ifdef ENABLE_LIBPNG
    else if (param.bg_format == "png") { }
#endif
#ifdef ENABLE_LIBJPEG
    else if (param.bg_format == "jpg") { }
#endif
#if ENABLE_SVG
    else if(param.bg_format == "svg") { }
#endif
    else
    {
        cerr << "Image format not supported: " << param.bg_format << endl;
        exit(EXIT_FAILURE);
    }

#if not ENABLE_SVG
    if(param.process_type3)
    {
        cerr << "process-type3 is enabled, however SVG support is not built in this version of pdf2htmlEX." << endl;
        exit(EXIT_FAILURE);
    }
#endif

    if((param.font_format == "ttf") && (param.external_hint_tool == ""))
    {
        cerr << "Warning: No hint tool is specified for truetype fonts, the result may be rendered poorly in some circumstances." << endl;
    }

    if (param.embed_image && (param.bg_format == "svg") && !param.svg_embed_bitmap)
    {
        cerr << "Warning: --svg-embed-bitmap is forced on because --embed-image is on, or the dumped bitmaps can't be loaded." << endl;
        param.svg_embed_bitmap = 1;
    }
}

int main(int argc, char **argv)
{
    // We need to adjust these directories before parsing the options.
#if defined(__MINGW32__)
    param.data_dir = get_exec_dir(argv[0]);
    param.tmp_dir  = get_tmp_dir();
#else
    char const* tmp = getenv("TMPDIR");
#ifdef P_tmpdir
    if (!tmp)
        tmp = P_tmpdir;
#endif
#ifdef _PATH_TMP
    if (!tmp)
        tmp = _PATH_TMP;
#endif
    if (!tmp)
        tmp = "/tmp";
    param.tmp_dir = string(tmp);
    param.data_dir = PDF2HTMLEX_DATA_PATH;
#endif

    if (getenv("APPDIR")) {
      // we are running inside an AppImage so we need to adjust the data_dir
      // however the user can supply some other absolute path later
      //
      param.data_dir = string(getenv("APPDIR")) + param.data_dir;
    }
    param.poppler_data_dir = param.data_dir + "/poppler";
    parse_options(argc, argv);
    check_param();

    //prepare the directories
    prepare_directories();

    if(param.debug)
        cerr << "temporary dir: " << (param.tmp_dir) << endl;

    try
    {
        create_directories(param.dest_dir);
    }
    catch (const string & s)
    {
        cerr << s << endl;
        exit(EXIT_FAILURE);
    }

    // setup the signal handler
    setupSignalHandler(argc, (const char**)argv,
      param.data_dir.c_str(),
      param.poppler_data_dir.c_str(),
      param.tmp_dir.c_str());

    bool finished = false;
    // read poppler config file
    globalParams = std::make_unique<GlobalParams>(
      !param.poppler_data_dir.empty() ? param.poppler_data_dir.c_str() : NULL
    );

    // open PDF file
    PDFDoc * doc = nullptr;
    try
    {
        {
            GooString * ownerPW = (param.owner_password == "") ? (nullptr) : (new GooString(param.owner_password.c_str()));
            GooString * userPW = (param.user_password == "") ? (nullptr) : (new GooString(param.user_password.c_str()));
            GooString fileName(param.input_filename.c_str());

            doc = PDFDocFactory().createPDFDoc(fileName, ownerPW, userPW);

            delete userPW;
            delete ownerPW;
        }

        if (!doc->isOk())
            throw "Cannot read the file";

        // check for copy permission
        if (!doc->okToCopy())
        {
            if (param.no_drm == 0)
                throw "Copying of text from this document is not allowed.";
            cerr << "Document has copy-protection bit set." << endl;
        }

        param.first_page =
          min<int>(max<int>(param.first_page, 1), doc->getNumPages());
        param.last_page =
          min<int>(max<int>(param.last_page, param.first_page),
                   doc->getNumPages());


        unique_ptr<HTMLRenderer>(new HTMLRenderer(argv[0], param))->process(doc);

        finished = true;
    }
    catch (const char * s)
    {
        cerr << "Error: " << s << endl;
    }
    catch (const string & s)
    {
        cerr << "Error: " << s << endl;
    }

    // clean up
    delete doc;
    globalParams.reset();

    // check for memory leaks
    // Poppler Object class (Object.h) no longer has memCheck
    //Object::memCheck(stderr);
    //gMemReport(stderr);

    exit(finished ? (EXIT_SUCCESS) : (EXIT_FAILURE));

    return 0;
}
